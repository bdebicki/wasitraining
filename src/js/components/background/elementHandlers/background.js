import LAYOUT from '../../../elementHandlers/layout';

const bgClassBuilders = {
	cover: LAYOUT.BG_COVER_ID,
	video: LAYOUT.BG_VIDEO_ID,
};
const BG = {
	COVER_CLASS: `${bgClassBuilders.cover}__cover`,
	VIDEO_CLASS: `${bgClassBuilders.video}__video`,
};

export { BG as default };
