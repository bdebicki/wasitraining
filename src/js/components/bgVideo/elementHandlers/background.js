import LAYOUT from '../../../elementHandlers/layout';

const bgClassBuilder = LAYOUT.BG_VIDEO_ID;
const BG = {
  VIDEO_CLASS: `${bgClassBuilder}__video`,
};

export { BG as default };
