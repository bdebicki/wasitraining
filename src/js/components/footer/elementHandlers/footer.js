const aboutClassBuilder = 'aboutSection';
const FOOTER = {
	COOKIES_CLASS: 'cookiesSection',
	ABOUT_SECTION_CLASS: aboutClassBuilder,
	COPYRIGHTS_CLASS: `${aboutClassBuilder}__copyrights`,
	ABOUT_LINK_CLASS: `${aboutClassBuilder}__aboutLink`,
};

export { FOOTER as default };
