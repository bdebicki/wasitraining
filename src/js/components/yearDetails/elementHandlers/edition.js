const editionClassBuilders = {
	details: 'editionDetails',
	lineup: 'shortLineup',
	headliner: 'headliner',
	headlines: 'headliners',
};
const EDITION = {
	EDITION_DETAILS_ID: editionClassBuilders.details,
	EDITION_DETAILS_YEAR_CLASS: `${editionClassBuilders.details}--edition`,
	FULL_NAME_CLASS: `${editionClassBuilders.details}__fullName`,
	PLACE_CLASS: `${editionClassBuilders.details}__place`,
	YEAR_CLASS: `${editionClassBuilders.details}__year`,
	DATES_CLASS: `${editionClassBuilders.details}__date`,
	LINEUP_CLASS: editionClassBuilders.lineup,
	LINEUP_EDITION_CLASS: `${editionClassBuilders.lineup}--edition`,
	HEADLINERS_CLASS: `${editionClassBuilders.lineup}__${editionClassBuilders.headlines}`,
	HEADLINER_CLASS: editionClassBuilders.headliner,
	HEADLINERS_DAY_CLASS: `${editionClassBuilders.headliner}--day`,
	LINEUP_LINK_CLASS: `${editionClassBuilders.lineup}__detailsLink`,
};

export {
	editionClassBuilders,
	EDITION,
};
