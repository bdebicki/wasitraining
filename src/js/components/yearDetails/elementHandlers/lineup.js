const lineupClassBuilder = {
  section: 'lineup',
  artists: 'lineupArtists',
  line: 'lineupArtists__line',
  lvl: 'lineupArtists__lvl',
  artist: 'lineupArtists__artist',
  slice: 'lineupArtists__slice',
  multipleArtists: 'lineupArtists__multipleArtists',
};
const LINEUP = {
  SECTION_ID: lineupClassBuilder.section,
  EDITION_CLASS: `${lineupClassBuilder.section}--edition`,
  ARTISTS_CLASS: lineupClassBuilder.artists,
  ARTISTS_EDITION_CLASS: `${lineupClassBuilder.artists}--edition`,
  ARTISTS_DAY_CLASS: `${lineupClassBuilder.artists}__day`,
  ARTISTS_NEW_LINE_ELEMENT_CLASS: `${lineupClassBuilder.artists}__newLine`,
  ARTISTS_SEPARATOR_ELEMENT_CLASS: `${lineupClassBuilder.artists}__separatorElement`,
  ARTISTS_LINE_CLASS: lineupClassBuilder.line,
  ARTISTS_LEVEL_CLASS: lineupClassBuilder.lvl,
  ARTIST_CLASS: lineupClassBuilder.artist,
  ARTIST_MULTIPLE_ARTISTS_CLASS: lineupClassBuilder.multipleArtists,
  ARTIST_MULTIPLE_ARTISTS_ARTIST_CLASS: `${lineupClassBuilder.multipleArtists}__artist`,
  ARTIST_SLICE_CLASS: `${lineupClassBuilder.slice}`,
};

export {
  lineupClassBuilder,
  LINEUP,
};
