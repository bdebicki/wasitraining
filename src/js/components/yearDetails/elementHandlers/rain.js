import globalClassBuilders from '../../../elementHandlers/globalClassBuilders';

const rainClassBuilders = {
  section: 'rainSection',
  info: 'rainInfo',
  header: 'rainHeader',
  details: 'rainDetails',
  detailsList: 'rainDetailsList',
};
const RAIN = {
  SECTION_ID: rainClassBuilders.section,
  INFO_CLASS: rainClassBuilders.info,
  INFO_YES_CLASS: `${rainClassBuilders.info}--yes`,
  INFO_NO_CLASS: `${rainClassBuilders.info}--no`,
  MASK_CLASS: `${rainClassBuilders.info}__mask`,
  YES_SHAPE_CLASS: `${rainClassBuilders.info}__maskShape--yes`,
  NO_SHAPE_CLASS: `${rainClassBuilders.info}__maskShape--no`,
  VIDEO_CLASS: `${rainClassBuilders.info}__video`,
  YES_VIDEO_CLASS: `${rainClassBuilders.info}__video--yes`,
  NO_VIDEO_CLASS: `${rainClassBuilders.info}__video--no`,
  HEADER_CLASS: rainClassBuilders.header,
  HEADER_HIDDEN_CLASS: `${rainClassBuilders.header}--${globalClassBuilders.hidden}`,
  HEADLINE_CLASS: `${rainClassBuilders.header}__headline`,
  DETAILS_LINK_CLASS: `${rainClassBuilders.details}__link`,
  CLOSE_DETAILS_CLASS: `${rainClassBuilders.details}__close`,
  DETAILS_ID: rainClassBuilders.details,
  DETAILS_CLASS: rainClassBuilders.details,
  DETAILS_LIST_CLASS: rainClassBuilders.detailsList,
  DETAILS_ITEM_CLASS: `${rainClassBuilders.detailsList}__item`,
  DETAILS_ITEM_DAY_CLASS: `${rainClassBuilders.detailsList}__day`,
  DETAILS_ITEM_RAIN_CLASS: `${rainClassBuilders.detailsList}__rain`,
};

export { RAIN as default };
