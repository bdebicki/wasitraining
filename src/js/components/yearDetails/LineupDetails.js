import Lineup from '../../classes/Lineup';
import {
	ARTIST_KEYS,
	ARTIST_SLICES_PROPS,
} from '../../enums/artist';
import { ARTIST_CANCELED } from '../../enums/content';
import LINEUP_LEVELS from '../../enums/lineupLevels';
import DIALOGBOX from '../../elementHandlers/dialogbox';
import EDITION from '../../elementHandlers/edition';
import { LINEUP } from '../../elementHandlers/lineup';
import LINK from '../../elementHandlers/link';
import * as dialogbox from '../../utils/addDialogbox';
import setIcon from '../../utils/setIcon';
import icons from '../../utils/iconsLibrary';
import getAlignClassName from './getAlignClassName';
import getLineupLvlClassName from './getLineupLvlClassName';
import getDecoratorClassName from './getDecoratorClassName';
import getSeparatorElementLvlClassName from './getSeparatorElementLvlClassName';
import getSliceDecoratorClassName from './getSliceDecoratorClassName';

const DIALOGBOX_HEADLINE_TEXT = 'Lineup';

function decoratorClassName(decorator) {
	if (Array.isArray(decorator)) {
		return decorator.map((decoratorType) => getDecoratorClassName(decoratorType));
	}

	return getDecoratorClassName(decorator);
}

export default class LineupDetails extends Lineup {
	static toggleLineup(e) {
		e.preventDefault();
		dialogbox.toggleDialogboxAction(`#${LINEUP.SECTION_ID}`, `#${EDITION.EDITION_DETAILS_ID}`);
	}

	static renderLineupLink() {
		const p = document.createElement('p');
		const a = document.createElement('a');

		p.classList.add(EDITION.LINEUP_LINK_CLASS);
		a.classList.add(LINK.BASIC_CLASS, LINK.INVERTED_STYLE_CLASS, LINK.SIZE_XS_CLASS, LINK.HAS_ICON_CLASS);
		a.href = '#lineup';
		a.textContent = 'see full lineup';
		a.appendChild(setIcon(icons.plus(), `${LINK.ICON_CLASS}`));
		a.addEventListener('click', LineupDetails.toggleLineup, null);

		p.appendChild(a);

		return p;
	}

	static getArtistSliceDecorator(artistName, artistDecorations) {
		const decorations = artistDecorations;
		const multipleDecorations = Array.isArray(decorations);
		const name = artistName;
		const decoratedName = document.createElement('template'); // can't use document fragment with innerHtml method
		let replacePattern;

		if (!multipleDecorations) {
			replacePattern = decorations[ARTIST_SLICES_PROPS.SLICE];
		} else {
			replacePattern = new RegExp(
				decorations.map((item) => item[ARTIST_SLICES_PROPS.SLICE]).join('|'), 'gi'
			);
		}

		decoratedName.innerHTML = name.replace(replacePattern, (matched) => {
			let sliceStyleClassName = '';
			const setStyleClassName = (styleList) => {
				if (!styleList[ARTIST_SLICES_PROPS.STYLE]) {
					return;
				}

				if (Array.isArray(styleList[ARTIST_SLICES_PROPS.STYLE])) {
					sliceStyleClassName = styleList[ARTIST_SLICES_PROPS.STYLE].map(
						(style) => getSliceDecoratorClassName(style)
					).join(' ');
				} else {
					sliceStyleClassName = getSliceDecoratorClassName(styleList[ARTIST_SLICES_PROPS.STYLE]);
				}
			};

			if (!multipleDecorations) {
				setStyleClassName(decorations);
			} else {
				decorations.forEach((decorationSlice) => {
					if (decorationSlice[ARTIST_SLICES_PROPS.SLICE] === matched) {
						setStyleClassName(decorationSlice);
					}
				});
			}
			return `<span class="${LINEUP.ARTIST_SLICE_CLASS} ${sliceStyleClassName}">${matched}</span>`;
		});

		return decoratedName.content;
	}

	static getArtistReplacementDecorator(artist, artistName, target) {
		const spanReplacement = document.createElement('span');
		const spanCanceled = document.createElement('span');
		const canceledArtistEl = target.querySelector(`li[data-canceled-artist='${artist[ARTIST_KEYS.REPLACEMENT]}']`);

		spanReplacement.textContent = artistName;
		spanReplacement.classList.add(LINEUP.ARTIST_MULTIPLE_ARTISTS_ARTIST_CLASS);
		spanCanceled.textContent = canceledArtistEl.textContent;
		spanCanceled.classList.add(LINEUP.ARTIST_CANCELED_CLASS, LINEUP.ARTIST_MULTIPLE_ARTISTS_ARTIST_CLASS);
		spanCanceled.title = ARTIST_CANCELED;

		// clean current existing canceled artist and put canceled artist with replacement
		canceledArtistEl.textContent = '';
		canceledArtistEl.removeAttribute('title');
		canceledArtistEl.classList.remove(LINEUP.ARTIST_CANCELED_CLASS, LINEUP.ARTIST_COLLAPSED_CLASS, LINEUP.ARTIST_EXPANDED_CLASS); // eslint-disable-line max-len
		canceledArtistEl.classList.add(LINEUP.ARTIST_MULTIPLE_ARTISTS_CLASS);
		canceledArtistEl.appendChild(spanCanceled);
		canceledArtistEl.appendChild(spanReplacement);
	}

	static getArtistClassNames(artist, artistLvl) {
		// TODO: move to external function
		const lvlClass = () => {
			if (artistLvl === LINEUP_LEVELS.HEADLINERS) {
				return `${LINEUP.ARTIST_CLASS}--headliner`;
			}

			return `${LINEUP.ARTIST_CLASS}--${artistLvl}`;
		};
		let classNames = [
			LINEUP.ARTIST_CLASS,
			lvlClass(),
			getAlignClassName(artist[ARTIST_KEYS.ALIGNED]),
			decoratorClassName(artist[ARTIST_KEYS.DECORATOR]),
			artist[ARTIST_KEYS.MARKED] ? LINEUP.ARTIST_MARKED_CLASS : null,
			artist[ARTIST_KEYS.MULTILINE] ? LINEUP.ARTIST_MULTILINE_CLASS : null,
			artist[ARTIST_KEYS.FIRST_ON_LINE] ? LINEUP.ARTIST_FIRST_ON_LINE_CLASS : null,
			artist[ARTIST_KEYS.LAST_ON_LINE] ? LINEUP.ARTIST_LAST_ON_LINE_CLASS : null,
			artist[ARTIST_KEYS.LAST_ON_DAY] ? LINEUP.ARTIST_LAST_ON_DAY_CLASS : null,
			artist[ARTIST_KEYS.REPLACEMENT] && typeof artist[ARTIST_KEYS.REPLACEMENT] !== 'string' ? LINEUP.ARTIST_REPLACEMENT_CLASS : null, // eslint-disable-line max-len
		];

		classNames = classNames
			.reduce((acc, className) => acc.concat(className), [])
			.filter((className) => { // clean null records on array
				if (className !== null) {
					return className;
				}

				return null;
			});

		return classNames;
	}

	static getArtistName(artist) {
		let artistName;

		if (artist[ARTIST_KEYS.ARTIST] && !artist[ARTIST_KEYS.DISPLAY_NAME]) {
			artistName = artist[ARTIST_KEYS.ARTIST];
		} else if (artist[ARTIST_KEYS.DISPLAY_NAME]) {
			artistName = artist[ARTIST_KEYS.DISPLAY_NAME];
		} else {
			artistName = artist;
		}

		return artistName;
	}

	static isArtistFirstOnLine(artist, target) {
		const newLine = document.createElement('li');

		newLine.classList.add(LINEUP.ARTISTS_NEW_LINE_ELEMENT_CLASS);

		target.appendChild(newLine);
	}

	static isArtistBreakLine(artist, target) {
		const previousEl = target.querySelector('li:last-child');

		previousEl.classList.add(LINEUP.ARTIST_NEXT_LINE_ARTIST_CLASS);
		previousEl.dataset[ARTIST_KEYS.NEXT_LINE_ARTIST] = artist[ARTIST_KEYS.SLICE_DECORATOR][ARTIST_SLICES_PROPS.SLICE]; // eslint-disable-line max-len
	}

	static isArtistCanceled(artist, artistName, target) {
		const targetEl = target;
		const span = document.createElement('span');

		span.classList.add(LINEUP.ARTIST_CANCELED_CLASS);
		span.textContent = artistName;
		targetEl.title = ARTIST_CANCELED;
		targetEl.dataset.canceledArtist = artistName;
		targetEl.appendChild(span);
	}

	static addSeparatorElement(target, lvl, element = 'li') {
		const separatorEl = document.createElement(element);

		separatorEl.classList.add(LINEUP.ARTISTS_SEPARATOR_ELEMENT_CLASS, getSeparatorElementLvlClassName(lvl));

		target.appendChild(separatorEl);
	}

	/* eslint-disable complexity */
	static decorateArtist(artist, target, index, lvl, separatorElement) {
		const fragment = document.createDocumentFragment();
		const li = document.createElement('li');
		const artistObj = artist;
		const artistName = LineupDetails.getArtistName(artistObj);
		const artistClassNames = LineupDetails.getArtistClassNames(artistObj, lvl);
		const hasSliceDecorator = artistObj[ARTIST_KEYS.SLICE_DECORATOR];
		const hasSeparatorElement = separatorElement;
		const isFirstOnLine = artistObj[ARTIST_KEYS.FIRST_ON_LINE];
		const isNewLine = artistObj[ARTIST_KEYS.NEW_LINE];
		const isBreakLine = artistObj[ARTIST_KEYS.BREAK_LINE];
		const isCanceled = artistObj[ARTIST_KEYS.CANCELED];
		const artistReplacement = artistObj[ARTIST_KEYS.REPLACEMENT];

		if (hasSeparatorElement && !isFirstOnLine && !artistObj[ARTIST_KEYS.SKIP_SEPARATOR]) {
			LineupDetails.addSeparatorElement(fragment, lvl);
		}

		if ((isFirstOnLine && index > 0) || isNewLine) {
			LineupDetails.isArtistFirstOnLine(artistObj, fragment);
		}

		if (isBreakLine) {
			LineupDetails.isArtistBreakLine(artistObj, target);
		}

		if (isCanceled) {
			LineupDetails.isArtistCanceled(artistObj, artistName, li);
		}

		if (hasSliceDecorator) {
			li.append(LineupDetails.getArtistSliceDecorator(artistName, artistObj[ARTIST_KEYS.SLICE_DECORATOR]));
		} else if (typeof artistReplacement === 'string') {
			LineupDetails.getArtistReplacementDecorator(artistObj, artistName, target);
		} else if (!isCanceled) {
			li.textContent = artistName;
		}

		li.classList.add(...artistClassNames);

		if (typeof artistReplacement !== 'string') {
			fragment.appendChild(li);
			target.appendChild(fragment);
		}
	}
	/* eslint-enable complexity */

	getLineupByType() {
		switch (this.mergeArtistsType) {
		case true:
			return this.decorateLineupByLevels();
		case 'mainByDaysAndMergeRest':
			return this.decorateMainByDaysAndMergeRest();
		case 'customLevels':
			return this.decorateLineupCustomLevelsByDays();
		default:
			return this.decorateLineupByDays();
		}
	}

	decorateLineupByLevels() {
		const fragment = document.createDocumentFragment();
		const { lineup, separatorElement } = this;

		Object.keys(lineup).forEach((lvl) => {
			const ul = document.createElement('ul');

			ul.classList.add(LINEUP.ARTISTS_LEVEL_CLASS, getLineupLvlClassName(lvl));

			lineup[lvl].forEach((artist, index) => {
				LineupDetails.decorateArtist(artist, ul, index, lvl, separatorElement);
			});

			fragment.appendChild(ul);
		});

		return fragment;
	}

	decorateMainByDaysAndMergeRest() {
		const fragment = document.createDocumentFragment();
		const { lineup, separatorElement } = this;

		Object.keys(lineup).forEach((section) => {
			if (section === LINEUP_LEVELS.DAILY_ARTISTS) {
				lineup[section].forEach((dailyArtists, dayIndex) => {
					const ul = document.createElement('ul');
					const dayCount = dayIndex + 1;

					ul.classList.add(
						LINEUP.ARTISTS_LEVEL_CLASS,
						getLineupLvlClassName(section),
						`${LINEUP.ARTISTS_LEVEL_CLASS}--day${dayCount}`
					);

					Object.keys(dailyArtists).forEach((dailyLvl) => {
						dailyArtists[dailyLvl].forEach((dailyArtist, dailyArtistIndex) => {
							LineupDetails.decorateArtist(dailyArtist, ul, dailyArtistIndex, dailyLvl, separatorElement);
						});
					});

					fragment.appendChild(ul);
				});
			} else if (section === LINEUP_LEVELS.ALL_OTHERS) {
				const ul = document.createElement('ul');

				ul.classList.add(LINEUP.ARTISTS_LEVEL_CLASS, getLineupLvlClassName(section));

				Object.keys(lineup[section]).forEach((lvl) => {
					lineup[section][lvl].forEach((artist, index) => {
						LineupDetails.decorateArtist(artist, ul, index, lvl, separatorElement);
					});
				});

				fragment.appendChild(ul);
			}
		});

		return fragment;
	}

	decorateLineupByDays() {
		const fragment = document.createDocumentFragment();
		const { lineup } = this;

		lineup.forEach((day, index) => {
			const section = document.createElement('section');
			const dayCount = index + 1;

			section.classList.add(LINEUP.ARTISTS_DAY_CLASS, `${LINEUP.ARTISTS_DAY_CLASS}--day${dayCount}`);

			Object.keys(day).forEach((lvl) => {
				const ul = document.createElement('ul');

				ul.classList.add(LINEUP.ARTISTS_LEVEL_CLASS, getLineupLvlClassName(lvl));

				day[lvl].forEach((artist, artistIndex) => {
					LineupDetails.decorateArtist(artist, ul, artistIndex, lvl);
				});

				section.appendChild(ul);
			});

			fragment.appendChild(section);
		});

		return fragment;
	}

	decorateLineupCustomLevelsByDays() {
		const fragment = document.createDocumentFragment();
		const { lineup, separatorElement } = this;
		const tempLineup = [];

		// merge all day artists in one array
		lineup.forEach((day, index) => {
			tempLineup.push([]); // create day key on tempLineup array

			Object.keys(day).forEach((level) => { // iterate on day
				day[level].forEach((artist) => { // iterate on level of specific day
					const artistObj = artist;
					const artistLine = artistObj[ARTIST_KEYS.LINE] - 1;
					artistObj[ARTIST_KEYS.LEVEL] = level;

					if (tempLineup[index][artistLine] === undefined) { // create line if doesn't exist
						// check does previous line exist; if not add empty arrays for lines above
						for (let i = 0; i <= artistLine; i++) {
							if (tempLineup[index][i] === undefined) {
								tempLineup[index].push([]);
							}
						}
					}

					tempLineup[index][artistLine].push(artistObj); // push artist into line
				});
			});
		});
		tempLineup.forEach((day, index) => {
			const section = document.createElement('section');
			const currentDay = index + 1;

			section.classList.add(LINEUP.ARTISTS_DAY_CLASS, `${LINEUP.ARTISTS_DAY_CLASS}--day${currentDay}`);

			day.forEach((line, lineIndex) => {
				if (line.length === 0) {
					return;
				}

				const ul = document.createElement('ul');
				const currentLine = lineIndex + 1;

				ul.classList.add(LINEUP.ARTISTS_LINE_CLASS, `${LINEUP.ARTISTS_LINE_CLASS}--line${currentLine}`);

				line.forEach((artist, artistIndex) => {
					LineupDetails.decorateArtist(artist, ul, artistIndex, artist[ARTIST_KEYS.LEVEL], separatorElement);
				});

				section.appendChild(ul);
			});

			fragment.appendChild(section);
		});

		return fragment;
	}

	updateLineupDetails() {
		const lineupContainer = document.querySelector(`#${LINEUP.SECTION_ID}`);
		const artistsContainer = document.querySelector(`.${LINEUP.ARTISTS_CLASS}`);
		const oldYear = lineupContainer.dataset.year;
		const newYear = this.editionYear;

		// eslint-disable-next-line max-len
		lineupContainer.querySelector(`.${DIALOGBOX.HEADLINE_CLASS}`).textContent = `${DIALOGBOX_HEADLINE_TEXT} ${newYear}`;
		lineupContainer.classList.remove(`${LINEUP.EDITION_CLASS}${oldYear}`);
		lineupContainer.classList.add(`${LINEUP.EDITION_CLASS}${newYear}`);
		lineupContainer.dataset.year = newYear;
		artistsContainer.textContent = '';
		artistsContainer.classList.remove(`${LINEUP.ARTISTS_EDITION_CLASS}${oldYear}`);
		artistsContainer.classList.add(`${LINEUP.ARTISTS_EDITION_CLASS}${newYear}`);
		artistsContainer.appendChild(this.getLineupByType());
	}

	render() {
		const section = document.createElement('section');
		const newYear = this.editionYear;

		const dialogboxLineup = dialogbox.addDialogbox({
			id: LINEUP.SECTION_ID,
			classNames: [`${LINEUP.EDITION_CLASS}${newYear}`],
			closeAction: dialogbox.toggleDialogboxWithInactive,
			dataAttr: [['year', `${newYear}`]],
			title: `${DIALOGBOX_HEADLINE_TEXT} ${newYear}`,
			content: section,
			closeTitle: 'hide lineup details',
		});

		section.classList.add(
			DIALOGBOX.CONTENT_CLASS,
			LINEUP.ARTISTS_CLASS,
			`${LINEUP.ARTISTS_EDITION_CLASS}${newYear}`
		);
		section.appendChild(this.getLineupByType());

		return dialogboxLineup;
	}
}
