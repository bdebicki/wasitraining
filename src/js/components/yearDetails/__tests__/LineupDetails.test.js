import * as artistsMock from '../../../../../tests/__mocks__/artists';
import editionCustomLevels from '../../../../../tests/__mocks__/edition-notSort-customLevels.json';
import editionMerged from '../../../../../tests/__mocks__/edition-customOrder-mergeArtists.json';
import editionMainByDaysAndMergeRest from '../../../../../tests/__mocks__/edition-customOrderExceptHeadliners-mainByDaysAndMergeRest.json';
import editionNotMerged from '../../../../../tests/__mocks__/edition-alphabeticalExceptHeadliners-noMergeArtists.json';
import LineupDetails from '../LineupDetails';
import { LINEUP, lineupClassBuilder } from '../elementHandlers/lineup';
import { ARTIST_KEYS } from '../../../enums/artist';
import EDITION from '../elementHandlers/edition';

describe('lineup details tests', () => {
	describe('single decorated artist (decorateArtist)', () => {
		// having
		let fragment;
		const lineup = new LineupDetails(editionCustomLevels);
		beforeEach(() => { fragment = document.createDocumentFragment(); });

		it('renders decorated artist element', () => {
			// when
			lineup.decorateArtist(artistsMock.artistObject, fragment);

			// then
			expect(fragment.querySelectorAll(`.${LINEUP.ARTIST_CLASS}`)).toHaveLength(1);
		});

		describe('artist additional classes', () => {
			it('add lvl class name', () => {
				// when
				lineup.decorateArtist(artistsMock.artistString, fragment, 1, 'headliners');

				// then
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--headliner`)).toBeTruthy();
			});

			it('add align class name', () => {
				// when
				lineup.decorateArtist(artistsMock.aligned, fragment);

				// then
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--rightAligned`)).toBeTruthy();
			});

			it('add decorator class name', () => {
				// when
				lineup.decorateArtist(artistsMock.decorator, fragment);

				// then
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--promoted`)).toBeTruthy();
			});

			it('add multiple decorators class name', () => {
				// when
				lineup.decorateArtist(artistsMock.multipleDecorator, fragment);

				// then
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--expanded`)).toBeTruthy();
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--wordExpandHeavy`)).toBeTruthy();
			});

			it('add marked class name', () => {
				// when
				lineup.decorateArtist(artistsMock.marked, fragment);

				// then
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--marked`)).toBeTruthy();
			});

			it('add multiline class name', () => {
				// when
				lineup.decorateArtist(artistsMock.multiline, fragment);

				// then
				expect(fragment.querySelector(`.${LINEUP.ARTIST_CLASS}--multiline`)).toBeTruthy();
			});
		});

		describe('artist with slice', () => {
			it('return artist with single slice decorator', () => {
				// when
				lineup.decorateArtist(artistsMock.slice, fragment);

				// then
				const slice = fragment.querySelectorAll(`.${LINEUP.ARTIST_SLICE_CLASS}`);
				expect(slice).toHaveLength(1);
				expect(slice[0].classList.contains(`${lineupClassBuilder.slice}--up`)).toBeTruthy();
				expect(slice[0].textContent).toBe('The');
			});

			it('return artist with multiple slice decorators', () => {
				// when
				lineup.decorateArtist(artistsMock.multipleSlice, fragment, 1, 'lvl1');

				// then
				expect(fragment).toMatchSnapshot();
			});

			it('return artist with multiple decorators for slice', () => {
				// when
				lineup.decorateArtist(artistsMock.multipleSliceStyles, fragment, 1, 'lvl1');

				// then
				const slice = fragment.querySelector(`.${LINEUP.ARTIST_SLICE_CLASS}`);
				expect(slice.classList.contains(`${LINEUP.ARTIST_SLICE_CLASS}--newLine`)).toBeTruthy();
				expect(slice.classList.contains(`${LINEUP.ARTIST_SLICE_CLASS}--indented`)).toBeTruthy();
			});
		});

		describe('artist name', () => {
			it('return artist from object', () => {
				// when
				lineup.decorateArtist(artistsMock.artistObject, fragment);

				// then
				expect(fragment.textContent).toBe('Pearl Jam');
			});

			it('return artist from string', () => {
				// when
				lineup.decorateArtist(artistsMock.artistString, fragment);

				// then
				expect(fragment.textContent).toBe('Pearl Jam');
			});

			it('return artist alternative display name', () => {
				// when
				lineup.decorateArtist(artistsMock.artistDisplayName, fragment);

				// then
				expect(fragment.textContent).toBe('Fisz Emade');
			});
		});

		describe('return canceled artist', () => {
			it('return clean canceled artist', () => {
				// when
				lineup.decorateArtist(artistsMock.canceled, fragment);

				// then
				expect(fragment.querySelector(`.${lineupClassBuilder.artist}--${ARTIST_KEYS.CANCELED}`)).toBeTruthy();
				expect(fragment.querySelector(`.${lineupClassBuilder.artist}`).title)
					.toBe('This artist\'s performance has been canceled');
			});

			it('return canceled artist with replacement', () => {
				// when
				artistsMock.canceledWithReplacement.forEach((artist) => lineup.decorateArtist(artist, fragment));

				// then
				const artist = fragment.querySelector(`.${lineupClassBuilder.artist}--${ARTIST_KEYS.CANCELED}`);
				expect(artist).toBeTruthy();
				expect(artist.title).toBe('This artist\'s performance has been canceled');
				expect(fragment.querySelectorAll(`.${LINEUP.ARTIST_MULTIPLE_ARTISTS_ARTIST_CLASS}`)).toHaveLength(2);
			});
		});

		describe('return artist with position options', () => {
			it('display artist first on line', () => {
				// when
				lineup.decorateArtist(artistsMock.firstOnLine, fragment);

				// then
				expect(fragment.querySelector(`.${lineupClassBuilder.artist}--${ARTIST_KEYS.FIRST_ON_LINE}`))
					.toBeTruthy();
			});

			it('display artist last on line', () => {
				// when
				lineup.decorateArtist(artistsMock.lastOnLine, fragment);

				// then
				expect(fragment.querySelector(`.${lineupClassBuilder.artist}--${ARTIST_KEYS.LAST_ON_LINE}`))
					.toBeTruthy();
			});

			it('display artist last on day', () => {
				// when
				lineup.decorateArtist(artistsMock.lastOnDay, fragment);

				// then
				expect(fragment.querySelector(`.${lineupClassBuilder.artist}--${ARTIST_KEYS.LAST_ON_DAY}`))
					.toBeTruthy();
			});

			it('display artist with new line', () => {
				// when
				lineup.decorateArtist(artistsMock.newLine, fragment);

				// then
				expect(fragment).toMatchSnapshot();
			});

			it('display artist broken on two lines', () => {
				// when
				artistsMock.twoLines.forEach((artist, index) => lineup.decorateArtist(artist, fragment, index, 'lvl1'));

				// then
				expect(fragment).toMatchSnapshot();
			});

			it('display artist with separator', () => {
				// when
				lineup.decorateArtist(artistsMock.artistString, fragment, 1, 'lvl1', true);

				// then
				expect(fragment).toMatchSnapshot();
			});

			it('display artist with turned off separator', () => {
				// when
				lineup.decorateArtist(artistsMock.turnedOffSeparator, fragment, 1, 'lvl1', true);

				// then
				expect(fragment).toMatchSnapshot();
			});
		});
	});

	describe('decorated lineup by type', () => {
		it('artists on lineup by levels', () => {
			// having
			const lineup = new LineupDetails(editionMerged);

			// when
			const decoratedLineup = lineup.getLineupByType();

			// then
			expect(decoratedLineup).toMatchSnapshot();
		});

		it('main artists by days and merge others', () => {
			// having
			const lineup = new LineupDetails(editionMainByDaysAndMergeRest);

			// when
			const decoratedLineup = lineup.getLineupByType();

			// then
			expect(decoratedLineup).toMatchSnapshot();
		});

		it('artist on custom levels', () => {
			// having
			const lineup = new LineupDetails(editionCustomLevels);

			// when
			const decoratedLineup = lineup.getLineupByType();

			// then
			expect(decoratedLineup).toMatchSnapshot();
		});

		it('artists on days', () => {
			// having
			const lineup = new LineupDetails(editionNotMerged);

			// when
			const decoratedLineup = lineup.getLineupByType();

			// then
			expect(decoratedLineup).toMatchSnapshot();
		});
	});

	describe('lineup dialogbox', () => {
		const lineup = new LineupDetails(editionCustomLevels);

		it('renders lineup dialogbox', () => {
			// having
			const basicLineup = new LineupDetails(editionMerged);

			// when
			document.body.appendChild(basicLineup.render());

			// then
			const dialogbox = document.getElementById(LINEUP.SECTION_ID);
			expect(dialogbox).toBeTruthy();
			expect(dialogbox.dataset.year).toBe('2007');
		});

		it('update lineup dialogbox', () => {
			// when
			lineup.updateLineupDetails();

			// then
			expect(document.getElementById(LINEUP.SECTION_ID).dataset.year).toBe('2004');
		});

		it('renders lineup link', () => {
			// having
			document.body.appendChild(LineupDetails.renderLineupLink());

			// then
			expect(document.querySelectorAll(`.${EDITION.LINEUP_LINK_CLASS}`)).toHaveLength(1);
		});

		it('toggle (show) lineup dialogbox visibility', () => {
		});
	});
});
