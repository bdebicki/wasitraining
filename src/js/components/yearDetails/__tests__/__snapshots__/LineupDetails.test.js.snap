// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lineup details tests single decorated artist (decorateArtist) artist with slice return artist with multiple slice decorators 1`] = `
<DocumentFragment>
  <li
    class="lineupArtists__artist lineupArtists__artist--undefined"
  >
    Zbigniew Wodeski 
    <span
      class="lineupArtists__slice lineupArtists__slice--down"
    >
      with
    </span>
     Mitch & Mitch 
    <span
      class="lineupArtists__slice lineupArtists__slice--multiline"
    >
      Orchestra and Choir
    </span>
  </li>
</DocumentFragment>
`;
